Run transactional callbacks on instances most likely to match DB state

cbothner opened the pull request:
When multiple instances of the same ActiveRecord model are used to persist changes within a transaction, Rails always runs `after_commit` callbacks on the first to save. This results in very unintuitive behavior:

```ruby
Product.transaction do
  Product.find(id).update!(title: "T-Shirt")
  Product.find(id).update!(description: "A cool T-shirt")
end
# When the transaction commits, `after_update_commit` callbacks are fired
# on the first instance, which does not have the new description.

Product.transaction do
  Product.find(id).update!(title: "T-Shirt")
  Product.find(id).destroy!
end
# When the transaction commits, `after_update_commit` callbacks are fired,
# even though the product is destroyed.
```

This PR implements a change to instead fire transactional callbacks on instances with internal state most likely to match the committed database state. 

- In general, run transactional callbacks on the last instance to save a given record within the transaction.
- There are two exceptions:
    - If the record is created within the transaction, then updated by another instance, `after_create_commit` callbacks will be run on the second instance. This is instead of the `after_update_commit` callbacks that would naively be run based on that instance’s state.
    - If the record is destroyed within the transaction, then `after_destroy_commit` callbacks will be fired on the last destroyed instance, even if a stale instance subsequently performed an update (which will have affected 0 rows).

Please see the added test cases for a more detailed statement of behavior.

cbothner committed:
Run transactional callbacks on instances most likely to match DB state

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:
```diff
@@ -188,6 +188,31 @@ def full_rollback?; true; end
       def open?; !closed?; end
+
+      private
+        def prepare_instances_to_run_callbacks_on(records)
```
We are now looping through records twice and building a hash even if we don't need to run commit callbacks. Can we merge both loops somehow?

cbothner replied:
When `@run_commit_callbacks` is true, this change is net zero on the number of hashes, since I was able to replace `already_run_callbacks` from inside of `commit_records`. It is an additional linear scan, but I don't think it's avoidable. For any given instance, in order to know whether to call `committed!` with a true or false value of `should_run_callbacks`, we need to have seen all other instances of the same record.

I will refactor to avoid the extra hash/iteration when `@run_commit_callbacks` is false.

rafaelfranca reviewed (commented):
This looks good to me. Let's add the flag to allow people to opt-in to it, but still makes this a framework default

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:
```diff
@@ -167,15 +167,17 @@ def before_commit_records
+
+        if @run_commit_callbacks
+          instances_to_run_callbacks_on = prepare_instances_to_run_callbacks_on(ite)
+          while record = ite.shift
```
```suggestion

          while record = ite.shift
```

cbothner commented:
@rafaelfranca I added the configuration flag, updated the framework defaults, adding documentation for the new flag, and added a changelog entry.

I considered whether a deprecation warning was necessary but based on the example of other such framework defaults I decided not to. We aren't really deprecating the old behavior if there's a configuration flag that lets an app developer opt in or out as they please. Let me know what you think.

CI is red but as far as I can tell it appears just to be flakiness.

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,33 @@
+    - In general, run transactional callbacks on the last instance to save a 
+      given record within the transaction.
+    - There are two exceptions:
+        - If the record is created within the transaction, then 
+          `after_create_commit` callbacks will be fired on the last instance to 
+          save, even if that instance performed an update.
```
Is this an exception? It is running the callback on the last instance to save.

cbothner replied:
The exceptional behavior is that we're running `after_create_commit`, not `after_update_commit` on that instance, even though from its perspective it just issued an update

https://github.com/rails/rails/blob/ad54b7a6733373b5426ceed6c4663f99e9781dbb/activerecord/test/cases/transaction_callbacks_test.rb#L791-L802

If you don't feel like that detail needs to be called out, then I'm happy to remove or reword.

rafaelfranca replied:
Ahhh, got it, that is not clear in the message. Can we update the message to show that difference?

cbothner replied:
Okay, I updated it. Is that more clear?

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,33 @@
+
+    When multiple Active Record instances change the same record within a 
+    transaction, Rails runs `after_commit` or `after_rollback` callbacks for 
+    only one of them. This adds an option to specify how Rails chooses which 
```
This part needs to be written in the past tense since CHANGELOGs entries are usually read after a release.

Something like:

```
`config.active_record.run_commit_callbacks_on_first_saved_instances_in_transaction` was added to specify how Rails chooses which instance receives the callbacks. The framework default was changed to defaults to the new
logic.
```

rafaelfranca commented on activerecord/test/cases/test_case.rb:
```diff
@@ -109,6 +109,18 @@ def with_automatic_scope_inversing(*reflections)
       end
     end
 
+    def with_run_commit_callbacks_on_first_saved_instances_in_transaction(value, model: ActiveRecord::Base)
```
This is only used in one test file, so let's move it there.

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_1.rb.tt:
```diff
@@ -28,3 +28,10 @@
 # as equal to an equivalent `Hash` by default.
 # Rails.application.config.action_controller.allow_deprecated_parameters_hash_equality = false
+
+# No longer run after_commit callbacks on the first of multiple ActiveRecord
```
```suggestion
# No longer run after_commit callbacks on the first of multiple Active Record
```

kshnurov commented:
@cbothner @rafaelfranca this change silently breaks `ActiveModel::Dirty` for `after_commit` callbacks. `saved_changes`, `saved_change_to_x?` and etc are pretty common in `after_commit`.

Depending on what else is happening in the transaction they'll either work as expected or will be completely ignored since the **last** object won't have the expected changes. A nightmare to debug and the only way to fix is to disable this behavior.

cbothner commented:
@kshnurov Anything that relies on instance state is unreliable when two instances are being used to write to the same row in the same transaction. No matter which instance we choose for callbacks, we'll miss something. Between the two options, I believe this is the better choice. Missing dirty state from the first instance to save does break callbacks that are conditional on `saved_changes`, but we provide the right column data to callbacks that don't rely on dirty state. Missing database state from the second to save breaks all callbacks that rely on that state, and sometimes results in the wrong lifecycle event being fired.

Really I think we should all try not to get into situations where we have multiple instances representing the same row. Passing around one instance will ensure no state is lost, and as a bonus it will save unnecessary `SELECT` statements.

kshnurov commented:
@cbothner well, both are bad, but the stale state is an [expected behaviour described in the docs](https://github.com/rails/rails/blob/main/guides/source/active_record_callbacks.md?plain=1#L598-L608). This PR didn't update them, which makes things even worse - people will expect it to work as before.

This one is a breaking change in Rails 7.1 that leads to silent hard-to-find bugs.
Fixing something while breaking the other thing is not a good solution.
Stale state is known, expected and could be easily fixed by a simple `.reload`. I don't see a workaround for this one though.

zzak commented:
@kshnurov Sorry to dismiss your comments here, but people make mistakes and responding to a PR that was merged over 18 months ago probably won't get the feedback you're looking for. I suggest if you can create a [reproduction](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#create-an-executable-test-case) to open a new issue. :pray:
